lib/TPP/Transforms/CombineXsmmPass.cpp

```diff
@@ -0,0 +1,193 @@
+          : *ldbDim;
+
+  auto ldcDim = xsmm::utils::getLeadingDim(memrefC);
+  if (failed(ldcDim))
+    return failure();
+
+  int64_t ldc = *ldcDim;
+
+  // If we are dealing with a BRGEMM we need to pass two extra dimensions:
+  // - strideA and strideB that represent the stride between different GEMM
+  // in BRGEMM.
+  int64_t strideA = lda * m;
+  int64_t strideB = ldb * k;
+  return DenseI64ArrayAttr::get(
+      rewriter.getContext(),
+      ArrayRef<int64_t>{m, n, k, lda, ldb, ldc, strideA, strideB});
+}
+
+static FailureOr<ArrayAttr> getBrgemmFlags(RewriterBase &rewriter,
+                                           xsmm::BrgemmOp opTy) {
+  auto memrefB = opTy.getOperand(2).getType().cast<MemRefType>();
+  auto flags =
+      dyn_cast<mlir::xsmm::BrgemmDispatchOp>(opTy.getOperand(0).getDefiningOp())
+          .getFlags();
+  for (auto flagItr : flags)
+    if (flagItr == xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                            mlir::xsmm::GemmFlags::BETA_0))
+      return failure();
+
+  xsmm::GemmFlagsAttr gemmFlag =
+      (vnni::utils::isInVnniLayout(vnni::utils::VnniOperandRank::BRGEMM_INS,
+                                   memrefB))
+          ? xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::VNNI_B)
+          : xsmm::GemmFlagsAttr::get(rewriter.getContext(),
+                                     xsmm::GemmFlags::NONE);
+  return rewriter.getArrayAttr(gemmFlag);
+}
+
+struct CombineXsmmOp : public OpRewritePattern<xsmm::BrgemmOp> {
+
+  using OpRewritePattern<xsmm::BrgemmOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(xsmm::BrgemmOp brgemmOp,
+                                PatternRewriter &rewriter) const override {
+    auto *output = brgemmOp.getOperand(3).getDefiningOp();
+    if (!output)
+      return failure();
+
+    // First, match the required fused ops
+    auto result = xsmm::utils::getFusedBrgemmSequenceFromProducer(output);
+    if (failed(result))
+      return failure();
+    auto fusedMatch = *result;
+    // TODO: Support BRGEMM + BINARY && BRGEMM + UNARY patterns
+    if (!fusedMatch.binaryOp || !fusedMatch.unaryOp)
+      return failure();
+    // Validate broadcast flags
+    auto unaryFlags =
+        xsmm::utils::getUnaryFlags(fusedMatch.unaryOp.getOperand(0).getType(),
+                                   fusedMatch.unaryOp.getOperand(2).getType());
+    if (unaryFlags != mlir::xsmm::UnaryFlags::BCAST_SCALAR &&
+        unaryFlags != mlir::xsmm::UnaryFlags::NONE)
+      return failure();
+
+    // TODO: Support more than just COL_0 BCAST
+    auto binaryFlags =
+        xsmm::utils::getBinaryFlags(fusedMatch.binaryOp.getOperand(1).getType(),
+                                    fusedMatch.binaryOp.getOperand(3).getType(),
+                                    mlir::xsmm::utils::OperandPos::LHS);
+    if (binaryFlags != mlir::xsmm::BinaryFlags::BCAST_COL_IN_0)
```

Comment:
Checked with @alheinecke, the behavior I describe above will change soon, so we should just have a simple conversion in the lowering:
```c++
  int binaryArg = 0;
  switch (binaryFlags) {
    case BinaryFlags::BCAST_COL_IN_0:
      binaryArg = 2;
      break;
    case BinaryFlags::BCAST_COL_IN_1:
      binaryArg = 3;
      binaryFlags = BinaryFlags::BCAST_COL_IN_0;
      break;
    default:
      return failure();
  }
```

Then lower down, you replace:
```c++
  invokeOperands.push_back(fusedMatch.binaryOp->getOperand(2));
```

with:
```c++
  invokeOperands.push_back(fusedMatch.binaryOp->getOperand(binaryArg));
```